### DISCUSSION TITLE
Buffer Size Problem - Audio Stream w/ EspNow and Ble - ESP32

### DISCUSSION THREAD

**Discussion by isadorabit** [pschatzmann/arduino-audio-tools#807] [Q&A] [answered]
Hi!  (Sorry for my bad english)
I'm working on this project:
1°ESP32: I take a mp3 file from SPIFFS and send via ESPNow to a second ESP32
2°ESP32: Receive the mp3 file and send to a bluetooth speaker
Everyting works well separately. But when the bluetooth speaker is connected with the 2°ESP32 and it is ready to receive data from the 1°ESP32,  occours a reset on the core 0 of 2°ESP32

Monitor Serial (2°ESP32):
"abort() was called at pc 0x401c52ef on core 0"
Backtrace: 0x4008f560:0x3ffd0240 0x4008f7dd:0x3ffd0260 0x401c52e7:0x3ffd0280 0x401c532e:0x3ffd02a0 0x401c4a07:0x3ffd02c0 0x401c4d3e:0x3ffd02e0 0x401c4aa5:0x3ffd0300 0x400d13b1:0x3ffd0320 0x400d2029:0x3ffd0340 0x400d2806:0x3ffd0360 0x401adc19:0x3ffd0390 0x400f3ba5:0x3ffd03b0 0x400f3a94:0x3ffd03e0 0x400ef746:0x3ffd0400 0x400ef8e9:0x3ffd0540 0x4010c8e1:0x3ffd0560 0x4010e46c:0x3ffd05a0 0x40090892:0x3ffd05d0

Using a backtrace decoder, i got this:

Decoding stack results:
0x4008f560: invoke_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/panic.c line 156
0x4008f7dd: abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp32/panic.c line 171
0x401c52e7: __cxxabiv1::__terminate(void (*)()) at /builds/idf/crosstool-NG/.build/src/gcc-5.2.0/libstdc++-v3/libsupc++/eh_terminate.cc line 47
0x401c532e: std::terminate() at /builds/idf/crosstool-NG/.build/src/gcc-5.2.0/libstdc++-v3/libsupc++/eh_terminate.cc line 57
0x401c4a07: __cxxabiv1::__cxa_throw(void*, std::type_info*, void (*)(void*)) at /builds/idf/crosstool-NG/.build/src/gcc-5.2.0/libstdc++-v3/libsupc++/eh_throw.cc line 87
0x401c4d3e: operator new(unsigned int) at /builds/idf/crosstool-NG/.build/src/gcc-5.2.0/libstdc++-v3/libsupc++/new_op.cc line 54
0x401c4aa5: operator new[](unsigned int) at /builds/idf/crosstool-NG/.build/src/gcc-5.2.0/libstdc++-v3/libsupc++/new_opv.cc line 32
0x400d13b1: audio_tools::RingBuffer ::resize(int) at C:\Users\isado\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/**Buffers.h line 297**
0x400d2029: audio_tools::RingBuffer ::RingBuffer(int) at C:\Users\isado\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/**Buffers.h line 234**
0x400d2806: audio_tools::ESPNowStream::default_recv_cb(unsigned char const*, unsigned char const*, int) at C:\Users\isado\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioLibs/**Communication.h line 275**
0x40090892: vPortTaskWrapper at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/port.c line 143

The three the highlighted lines point to the following code lines:

Buffers.h line 297  --> _aucBuffer = new T[max_size];
Buffers.h line 234 -->  RingBuffer(int size) { resize(size); } 
Communication.h line 275 --> p_buffer = new RingBuffer<uint8_t>(cfg.buffer_size * cfg.buffer_count);

This is all related to the buffer size, so I think this is the problem.

How can I solve this?
Is there a way to reduce the size of buffer?
I'm wondering if i can send the file in smaller pieces of data...

**Comment by pschatzmann** [comment, marked as answer]
I agree, you are running out of RAM!
Why don't you just decrease the buffer size of buffer count used by esp-now ?
Did you ever look at the available configuration parameters ?

You should also try to optimize the I2S buffer parameters...

**Comment by jajito** [comment]
Hi.
Have in mind that the Bluetooth and WIFI ( I suppose that esp-now too) share the same radio module.
I tried long ago to use both of them at the same time, the bluetooth sink should download the cover of the song, and couldn't make it work together. 
If the ESP32 Bluetooth was active, once it tried to use the wifi to download the cover or any other information from the web: reset.
(https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/coexist.html)

Good luck and if you get it to work let us know!
